// @slack/bolt/dist/receivers/SocketModeReceiver.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable @typescript-eslint/no-explicit-any */
const socket_mode_1 = require("@slack/socket-mode");
const http_1 = require("http");
const logger_1 = require("@slack/logger");
const oauth_1 = require("@slack/oauth");
const render_html_for_install_path_1 = __importDefault(require("./render-html-for-install-path"));
const custom_routes_1 = require("./custom-routes");
const verify_redirect_opts_1 = require("./verify-redirect-opts");
/**
 * Receives Events, Slash Commands, and Actions of a web socket connection
 */
class SocketModeReceiver {
    constructor({ appToken, logger = undefined, logLevel = logger_1.LogLevel.INFO, clientId = undefined, clientSecret = undefined, stateSecret = undefined, redirectUri = undefined, installationStore = undefined, scopes = undefined, installerOptions = {}, customRoutes = [], customPropertiesExtractor = (_args) => ({}), }) {
        var _a;
        this.installer = undefined;
        this.client = new socket_mode_1.SocketModeClient({
            appToken,
            logLevel,
            logger,
            clientOptions: installerOptions.clientOptions,
        });
        this.logger = logger !== null && logger !== void 0 ? logger : (() => {
            const defaultLogger = new logger_1.ConsoleLogger();
            defaultLogger.setLevel(logLevel);
            return defaultLogger;
        })();
        this.routes = (0, custom_routes_1.prepareRoutes)(customRoutes);
        // Verify redirect options if supplied, throws coded error if invalid
        (0, verify_redirect_opts_1.verifyRedirectOpts)({ redirectUri, redirectUriPath: installerOptions.redirectUriPath });
        if (clientId !== undefined &&
            clientSecret !== undefined &&
            (installerOptions.stateVerification === false || // state store not needed
                stateSecret !== undefined ||
                installerOptions.stateStore !== undefined) // user provided state store
        ) {
            this.installer = new oauth_1.InstallProvider({
                clientId,
                clientSecret,
                stateSecret,
                installationStore,
                logLevel,
                logger,
                stateStore: installerOptions.stateStore,
                stateVerification: installerOptions.stateVerification,
                authVersion: (_a = installerOptions.authVersion) !== null && _a !== void 0 ? _a : 'v2',
                clientOptions: installerOptions.clientOptions,
                authorizationUrl: installerOptions.authorizationUrl,
            });
        }
        // Add OAuth and/or custom routes to receiver
        if (this.installer !== undefined || customRoutes.length) {
            // use default or passed in redirect path
            const redirectUriPath = installerOptions.redirectUriPath === undefined ? '/slack/oauth_redirect' : installerOptions.redirectUriPath;
            // use default or passed in installPath
            const installPath = installerOptions.installPath === undefined ? '/slack/install' : installerOptions.installPath;
            const directInstallEnabled = installerOptions.directInstall !== undefined && installerOptions.directInstall;
            this.httpServerPort = installerOptions.port === undefined ? 3000 : installerOptions.port;
            this.httpServer = (0, http_1.createServer)(async (req, res) => {
                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                const method = req.method.toUpperCase();
                // Handle OAuth-related requests
                if (this.installer) {
                    // create install url options
                    const installUrlOptions = {
                        metadata: installerOptions.metadata,
                        scopes: scopes !== null && scopes !== void 0 ? scopes : [],
                        userScopes: installerOptions.userScopes,
                        redirectUri,
                    };
                    // Installation has been initiated
                    if (req.url && req.url.startsWith(redirectUriPath)) {
                        const { stateVerification, callbackOptions } = installerOptions;
                        if (stateVerification === false) {
                            // if stateVerification is disabled make install options available to handler
                            // since they won't be encoded in the state param of the generated url
                            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                            await this.installer.handleCallback(req, res, callbackOptions, installUrlOptions);
                        }
                        else {
                            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                            await this.installer.handleCallback(req, res, callbackOptions);
                        }
                        return;
                    }
                    // Visiting the installation endpoint
                    if (req.url && req.url.startsWith(installPath)) {
                        const { stateVerification, renderHtmlForInstallPath } = installerOptions;
                        try {
                            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                            const url = await this.installer.generateInstallUrl(installUrlOptions, stateVerification);
                            if (directInstallEnabled) {
                                res.writeHead(302, { Location: url });
                                res.end('');
                            }
                            else {
                                res.writeHead(200, {});
                                const renderHtml = renderHtmlForInstallPath !== undefined ?
                                    renderHtmlForInstallPath :
                                    render_html_for_install_path_1.default;
                                res.end(renderHtml(url));
                                return;
                            }
                        }
                        catch (err) {
                            const e = err;
                            throw new Error(e);
                        }
                    }
                }
                // Handle request for custom routes
                if (customRoutes.length && req.url) {
                    const match = this.routes[req.url] && this.routes[req.url][method] !== undefined;
                    if (match) {
                        this.routes[req.url][method](req, res);
                        return;
                    }
                }
                this.logger.info(`An unhandled HTTP request (${req.method}) made to ${req.url} was ignored`);
                res.writeHead(404, {});
                res.end();
            });
            this.logger.debug(`Listening for HTTP requests on port ${this.httpServerPort}`);
            if (this.installer) {
                this.logger.debug(`Go to http://localhost:${this.httpServerPort}${installPath} to initiate OAuth flow`);
            }
        }
        this.client.on('slack_event', async (args) => {
            var _a;
            const { ack, body, retry_num, retry_reason } = args;
            const event = {
                body,
                ack,
                retryNum: retry_num,
                retryReason: retry_reason,
                customProperties: customPropertiesExtractor(args),
            };
            await ((_a = this.app) === null || _a === void 0 ? void 0 : _a.processEvent(event));
        });
    }
    init(app) {
        this.app = app;
    }
    start() {
        if (this.httpServer !== undefined) {
            // This HTTP server is only for the OAuth flow support
            this.httpServer.listen(this.httpServerPort);
        }
        // start socket mode client
        return this.client.start();
    }
    stop() {
        if (this.httpServer !== undefined) {
            // This HTTP server is only for the OAuth flow support
            this.httpServer.close((error) => {
                this.logger.error(`Failed to shutdown the HTTP server for OAuth flow: ${error}`);
            });
        }
        return new Promise((resolve, reject) => {
            try {
                this.client.disconnect();
                resolve();
            }
            catch (error) {
                reject(error);
            }
        });
    }
}
exports.default = SocketModeReceiver;
